From 7bf56754796a24ea8777492351777f55b751bcb8 Mon Sep 17 00:00:00 2001
From: linmin <linmin@eswincomputing.com>
Date: Sat, 8 Jun 2024 15:39:50 +0800
Subject: [PATCH 080/109] WIN2030-15279:feat(console):Add eeprom write protect
 cmd

Changelogs:
1.Add eepromwp-s <1 or 0> ,i.e enable or disable eeprom write protect.
  This is only for burning eeprom on the production line and not
  available for users.So this command will not be shown with help command.
2.Add pwrlast-g command to get the som power last state from eeprom.
  It is used for test to determine whether the power off of the som is
  expected.

Change-Id: I9ed59dc65fc5e4988de918c3ae4a397039a146e7
---
 Core/Src/console.c | 68 ++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 68 insertions(+)

diff --git a/Core/Src/console.c b/Core/Src/console.c
index 0d52437..7c4db14 100644
--- a/Core/Src/console.c
+++ b/Core/Src/console.c
@@ -106,6 +106,12 @@ static BaseType_t prvCommandSomPwrLostResumeGet(char *pcWriteBuffer, size_t xWri
 // set the power lost resume attribute: enable or disable
 static BaseType_t prvCommandSomPwrLostResumeSet(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString);
 
+// get the power last state from eeprom: 1 on , 0 off
+static BaseType_t prvCommandSomPwrLastStateGet(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString);
+
+// eeprom write protect pin set
+static BaseType_t prvCommandCBWpEEPROM(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString);
+
 /**
 *   @brief  This function is executed in case of error occurrence.
 *   @retval None
@@ -281,6 +287,18 @@ static const CLI_Command_Definition_t xCommands[] =
         prvCommandSomPwrLostResumeSet,
         1
     },
+    {
+        "pwrlast-g",
+        "\r\npwrlast-g: Get the som power last state. 1 or 0.\r\n",
+        prvCommandSomPwrLastStateGet,
+        0
+    },
+    {
+        "eepromwp-s",
+        "",
+        prvCommandCBWpEEPROM,
+        1
+    },
     {
        "echo",
        "\r\necho <string to echo>\r\n",
@@ -1302,6 +1320,56 @@ static BaseType_t prvCommandSomPwrLostResumeSet(char *pcWriteBuffer, size_t xWri
     return pdFALSE;
 }
 
+
+/**
+* @brief Get the som power last state from eeprom: 1 on or 0 off
+* @param *pcWriteBuffer FreeRTOS CLI write buffer.
+* @param xWriteBufferLen Length of write buffer.
+* @param *pcCommandString pointer to the command name.
+* @retval FreeRTOS status
+*/
+static BaseType_t prvCommandSomPwrLastStateGet(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString)
+{
+    int pwr_last_state = 0;
+
+    /* pwr last state test */
+    es_get_som_pwr_last_state(&pwr_last_state);
+
+    snprintf(pcWriteBuffer, xWriteBufferLen, "pwr_last_state: %d\r\n", pwr_last_state);
+
+    return pdFALSE;
+}
+
+
+/**
+* @brief CarrierBoard EEPROM Write protect eeprom or disable write protect
+* @param *pcWriteBuffer FreeRTOS CLI write buffer.
+* @param xWriteBufferLen Length of write buffer.
+* @param *pcCommandString pointer to the command name.
+* @retval FreeRTOS status
+*/
+static BaseType_t prvCommandCBWpEEPROM(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString)
+{
+    const char *cEEPROM_WP;
+    BaseType_t xParamLen;
+    uint8_t WP_En;
+
+    cEEPROM_WP = FreeRTOS_CLIGetParameter(pcCommandString, 1, &xParamLen);
+    WP_En = atoi(cEEPROM_WP);
+
+    /* change write protect pin value */
+    if (!WP_En) {
+        HAL_GPIO_WritePin(EEPROM_WP_GPIO_Port, EEPROM_WP_Pin, GPIO_PIN_RESET);
+        snprintf(pcWriteBuffer, xWriteBufferLen, "EEPROM Write protect disabled!\r\n");
+    }
+    else {
+        HAL_GPIO_WritePin(EEPROM_WP_GPIO_Port, EEPROM_WP_Pin, GPIO_PIN_SET);
+        snprintf(pcWriteBuffer, xWriteBufferLen, "EEPROM Write protect enabled!\r\n");
+    }
+
+    return pdFALSE;
+}
+
 /**
 * @brief Reads from UART RX buffer. Reads one bye at the time.
 * @param *cReadChar pointer to where data will be stored.
-- 
2.25.1

